src/layers/domain/core/device_key/v2.py

```diff
@@ -0,0 +1,17 @@
+from domain.core.base import BaseModel
+from domain.core.device_key.v1 import DeviceKeyType, validate_key
+from pydantic import validator
+
+
+class DeviceKey(BaseModel):
+    """
+    A Device Key is a secondary way of indexing / retrieving Devices
+    """
+
+    key_type: DeviceKeyType
+    key_value: str
+
+    @validator("key_value", check_fields=True)
+    def validate_key(key_value: str, values: dict):
+        key_type: DeviceKeyType = values.get("key_type")
+        return validate_key(key=key_value, type=key_type)
```

Comment:
Refactor the validate_key function outside of this too?


```python
def validate_key(key_value: str, key_type: DeviceKeyType):
    if key_type and key_type.pattern.match(key) is None:
        raise InvalidDeviceKeyError(
            f"Key '{key_value}' does not match the expected "
            f"pattern '{key_type.pattern.pattern}' associated with "
            f"key type '{key_type}'"
        )
    return key_value
```

Reply:
yeah wasn't totally sure which side of the fence i sat on this one. Reason to not the private-ish internal function is so that the behaviour between versions is coupled without duplicated code. Feels like it would be easy to introduce bugs between versions if we duplicated too much